File: expectanim/src/main/java/com/github/florent37/expectanim/core/position/PositionAnimExpectation.java
package com.github.florent37.expectanim.core.position;

import android.content.Context;
import android.support.annotation.DimenRes;
import android.support.annotation.Nullable;
import android.util.TypedValue;
import android.view.View;

import com.github.florent37.expectanim.core.AnimExpectation;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by florentchampigny on 17/02/2017.
 */

public abstract class PositionAnimExpectation extends AnimExpectation {
    public abstract Float getCalculatedValueX(View viewToMove);
    public abstract Float getCalculatedValueY(View viewToMove);

    private boolean isForPositionY;
    private boolean isForPositionX;
    private boolean isForTranslationX;
    private boolean isForTranslationY;
    private float margin;

    @Nullable
    private Integer marginRes;

    @Nullable
    private Float marginDp;

    public boolean isForPositionY() {
        return isForPositionY;
    }

    public boolean isForPositionX() {
        return isForPositionX;
    }

    public boolean isForTranslationX() {
        return isForTranslationX;
    }

    public boolean isForTranslationY() {
        return isForTranslationY;
    }

    protected void setForPositionY(boolean forPositionY) {
        isForPositionY = forPositionY;
    }

    protected void setForPositionX(boolean forPositionX) {
        isForPositionX = forPositionX;
    }

    protected void setForTranslationX(boolean forTranslationX) {
        isForTranslationX = forTranslationX;
    }

    protected void setForTranslationY(boolean forTranslationY) {
        isForTranslationY = forTranslationY;
    }

    public float getMargin(View view) {
        if(marginRes != null){
            margin = view.getContext().getResources().getDimension(marginRes);
        } else if(marginDp != null){
            margin = dpToPx(view.getContext(), marginDp);
        }
        return margin;
    }

    public static float dpToPx(Context context, float dp){
        return TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, dp, context.getResources().getDisplayMetrics());
    }

    public PositionAnimExpectation withMargin(float margin){
        this.margin = margin;
        return this;
    }

    public PositionAnimExpectation withMarginDimen(@DimenRes int marginRes){
        this.marginRes = marginRes;
        return this;
    }

    public PositionAnimExpectation withMarginDp(float marginDp){
        this.marginDp = marginDp;
        return this;
    }

}

File: expectanim/src/main/java/com/github/florent37/expectanim/core/position/PositionAnimExpectation.kt
package com.github.florent37.expectanim.core.position

import android.content.Context
import android.support.annotation.DimenRes
import android.util.TypedValue
import android.view.View

import com.github.florent37.expectanim.core.AnimExpectation

import java.util.ArrayList

abstract class PositionAnimExpectation : AnimExpectation() {

    var isForPositionY: Boolean = false
        protected set
    var isForPositionX: Boolean = false
        protected set
    var isForTranslationX: Boolean = false
        protected set
    var isForTranslationY: Boolean = false
        protected set

    var margin: Float? = null
    var marginRes: Int? = null
    var marginDp: Float? = null

    abstract fun getCalculatedValueX(viewToMove: View): Float?
    abstract fun getCalculatedValueY(viewToMove: View): Float?

    fun getMargin(view: View): Float {
        return when {
            this.margin != null -> this.margin!!
            marginRes != null -> view.context.resources.getDimension(marginRes!!)
            marginDp != null -> dpToPx(view.context, marginDp!!)
            else -> 0f
        }
    }

    companion object {
        fun dpToPx(context: Context, dp: Float): Float {
            return TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, dp, context.resources.displayMetrics)
        }
    }

}
